americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+from __future__ import division
+from os import mkdir
+from os.path import exists
+
+import numpy as np
+import pandas as pd
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import skbio
+
+from scipy.stats import kruskal
+from skbio.stats.power import _check_strs
+from statsmodels.sandbox.stats.multicomp import multipletests
+
+__author__ = "Justine Debelius"
+__copyright__ = "Copyright 2015, The American Gut Project"
+__credits__ = ["Justine Debelius"]
+__license__ = "BSD"
+__version__ = "unversioned"
+__maintainer__ = "Justine Debelius"
+__email__ = "Justine.Debelius@colorado.edu"
+
+
+def check_dir(dir_):
+    """Creates the specified directory if it does not exist
+
+    Parameters
+    ----------
+    dir : str
+        the directory to be checked
+    """
+
+    if not exists(dir_):
+        mkdir(dir_)
+
+
+def pad_index(df, index_col='#SampleID', nzeros=9):
+    """Adds zeros to the sample ID strings
+
+    Parameters
+    ----------
+    df : dataframe
+        the data frame without an index column
+    index_col : {#SampleID, str}
+        the name of the column containing the index data
+    n_zeros : {9, int}
+        the number of zeros to add before the string
+
+    Returns
+    -------
+    df : dataframe
+        the dataframe with an appropriate index column
+    """
+
+    # Gets the sample IDs
+    samples = df[index_col].values
+    new_samples = []
+
+    # Pads the zeros on the id
+    for samp in samples:
+        if not isinstance(samp, str):
+            samp = str(samp)
+        splits = samp.split('.')
+        first_clean = [splits[0].zfill(nzeros)]
+        first_clean.extend(splits[1:])
+        new_samples.append('.'.join(first_clean))
+
+    # Sets the column as the index
+    df.index = new_samples
+    del df[index_col]
+
+    # Returns the dataframe
+    return df
+
+
+def boxplot(vecs, ax=None, notch=True, interval=0.5, boxplot_props={},
```

Comment:
use `None` and test instead of `{}` for defaults. The reason is that `{}` is eval'd once when the function is defined, so if it gets modified, the modification remains. For instance:

```python
>>> def foo(a, b={}):
...   b[a] = a
...   print b
...
>>> foo(10)
{10: 10}
>>> foo(100)
{10: 10, 100: 100}
>>> foo(1000)
{1000: 1000, 10: 10, 100: 100}
```